### DISCUSSION TITLE
Failed restore causes later updates to fail

### DISCUSSION THREAD

**Discussion by goodevilgenius** [owen-it/laravel-auditing#793] [Q&A]
| Q                 | A
| ----------------- | ---
| Bug?              | yes
| New Feature?      | no
| Framework         | Lumen
| Framework version | 5.8.13
| Package version   | 9.3.2
| PHP version       | 7.2.18

#### Actual Behaviour

Update audit event does not occur if a previous restore failed.

#### Expected Behaviour

Update should happen regardless of previous restore.

#### Steps to Reproduce

Image code along the following lines:

```php
$model1 = Model::onlyTrashed()->find(23); // returns trashed model
$model1->restore();
// Some restoring observer (which runs after the AuditingObserver) returns false,
// causing the restore to fail

$model2 = Model->find(42);
$model2->update(['key' => 'value']);
```

Because `$model1->restore` failed, the `restored` model event was never fired. Because of this, `AuditableObserver::$restoring` is still set to true.

Because `AuditableObserver::$restoring` is still `true`, `$model2->update` doesn't trigger the `updated` audit event.

#### Possible Solutions

I worked around this locally by modifying my model to have a `$restoring` property initialized to `false`. I then created my own `restore` method like this:

```php
public function restore()
{
    $this->restoring = true;
    $restored = false;

    try {
        $restored = parent::restore();
    } finally {
        $this->restoring = false;
    }

    return $restored;
}


/**
 * Indicates if the document is being restored.
 *
 * @return bool
 */
public function isRestoring(): bool
{
    return $this->restoring;
}
```

I then extended `AuditableObserver` and overrode the `updated` method like this:

```php
public function updated(Auditable $model)
{
    if (!$model->isRestoring()) {
        static::$restoring = false;
        parent::updated($model);
    }
}
```

And then, of course, overrode the `bootAuditable` method to observe my new `AuditableObserver`.

**Comment by parallels999** [comment]
Hi, If you know how to solve it please make a PR with some tests to avoid breaking that on the future, thanks
